File: clients/python/outbox_client.py
import json
from urllib.error import HTTPError
from urllib.request import Request, urlopen


class OutboxClient:
    def __init__(self, base_url, token, timeout=30):
        self.base_url = base_url.rstrip("/")
        self.token = token
        self.timeout = timeout

    def _post(self, path, payload):
        request = Request(
            self.base_url + path,
            data=json.dumps(payload).encode("utf-8"),
            headers={
                "Authorization": f"Bearer {self.token}",
                "Content-Type": "application/json",
            },
            method="POST",
        )
        try:
            with urlopen(request, timeout=self.timeout) as response:
                return json.loads(response.read().decode("utf-8"))
        except HTTPError as exc:
            detail = exc.read().decode("utf-8", errors="replace")
            raise RuntimeError(f"Outbox API returned HTTP {exc.code}: {detail}") from exc

    def claim(self, worker_id, channels, providers, max_items=1, wait_seconds=20):
        return self._post(
            "/v1/deliveries/claim",
            {
                "worker_id": worker_id,
                "channels": channels,
                "providers": providers,
                "max_items": max_items,
                "wait_seconds": wait_seconds,
            },
        )["items"]

    def heartbeat(self, item, worker_id):
        return self._post(
            f"/v1/deliveries/{item['delivery_id']}/heartbeat",
            {"worker_id": worker_id, "lease_token": item["lease_token"]},
        )

    def complete(self, item, worker_id, provider_message_id=None, provider_thread_id=None):
        return self._post(
            f"/v1/deliveries/{item['delivery_id']}/complete",
            {
                "worker_id": worker_id,
                "lease_token": item["lease_token"],
                "provider_message_id": provider_message_id,
                "provider_thread_id": provider_thread_id,
            },
        )

    def fail(self, item, worker_id, result, error_code, error_message=""):
        return self._post(
            f"/v1/deliveries/{item['delivery_id']}/fail",
            {
                "worker_id": worker_id,
                "lease_token": item["lease_token"],
                "result": result,
                "error_code": error_code,
                "error_message": error_message,
            },
        )

File: clients/typescript/outbox-client.ts
export type Delivery = {
  delivery_id: string;
  message_version_id: string;
  lease_token: string;
  lease_expires_at: string | null;
  schema_version: "1.0";
  idempotency_key: string;
  channel: "email" | "linkedin";
  provider: string;
  recipient: string;
  sender_account_ref?: string | null;
  payload: Record<string, unknown>;
  metadata: { lead_id?: string; attempt_no: number };
};

export class OutboxClient {
  constructor(
    private readonly baseUrl: string,
    private readonly token: string,
  ) {}

  private async post<T>(path: string, body: unknown): Promise<T> {
    const response = await fetch(`${this.baseUrl.replace(/\/$/, "")}${path}`, {
      method: "POST",
      headers: {
        Authorization: `Bearer ${this.token}`,
        "Content-Type": "application/json",
      },
      body: JSON.stringify(body),
    });
    if (!response.ok) {
      throw new Error(`Outbox API ${response.status}: ${await response.text()}`);
    }
    return response.json() as Promise<T>;
  }

  async claim(
    workerId: string,
    channels: Array<"email" | "linkedin">,
    providers: string[],
    maxItems = 1,
    waitSeconds = 20,
  ): Promise<Delivery[]> {
    const result = await this.post<{ items: Delivery[] }>("/v1/deliveries/claim", {
      worker_id: workerId,
      channels,
      providers,
      max_items: maxItems,
      wait_seconds: waitSeconds,
    });
    return result.items;
  }

  heartbeat(item: Delivery, workerId: string) {
    return this.post(`/v1/deliveries/${item.delivery_id}/heartbeat`, {
      worker_id: workerId,
      lease_token: item.lease_token,
    });
  }

  complete(
    item: Delivery,
    workerId: string,
  ) {
    return this.post(`/v1/deliveries/${item.delivery_id}/complete`, {
      worker_id: workerId,
    });
  }

  fail(
    item: Delivery,
    workerId: string,
  ) {
    return this.post(`/v1/deliveries/${item.delivery_id}/fail`, {
      worker_id: workerId,
    });
  }
}
